### DISCUSSION TITLE
The problem with network interfaces

### DISCUSSION THREAD

**Discussion by Chr-Sergey** [lae/ansible-role-proxmox#175] [Q&A]
Each node has an external and internal network (cluster network). Proxmox is installed on the nodes, the cluster is deployed. But the communication of the cluster takes place on an external  IP. Need an internal IP.

Nodes internal IP:
192.168.56.25
192.168.56.26
192.168.56.27
How to use internal ip  for cluster network.

**Comment by lae** [comment]
You can use the `addr1` role variable below to deploy PVE on a separate (private) cluster network, or override the `addr0` role variable with your private network's IP address.

```yaml
pve_cluster_addr0: "{{ ansible_default_ipv4.address }}"
pve_cluster_addr1: "another interface's IP address or hostname"
```

e.g. you could use something like `pve_cluster_addr0: "{{ ansible_eth1.ipv4.address }}"` to configure PVE to deploy on the private network interface `eth1`.

P.S. You may also want to read the section on Role Variables in the README: https://github.com/lae/ansible-role-proxmox#role-variables

**Comment by Chr-Sergey** [comment]
@lae Thanks.
But,  if use pve_cluster_addr0: "{{ ansible_interfaces.enp0s8.ipv4_address }}",  I have error:

.....
'list object' has no attribute 'enp0s8'\n\nThe error appears to be in '/mnt/d/proxmox/test/ansible-role-proxmox/tasks/load_variables.yml'
.....


fatal: [192.168.15.152]: FAILED! => {
    "msg": "The conditional check 'pve_cluster_addr0 is not defined' failed. The error was: error while evaluating conditional (pve_cluster_addr0 is not defined): {{ ansible_interfaces.enp0s8.ipv4.address }}: 'list object' has no attribute 'enp0s8'\n\nThe error appears to be in '/mnt/d/proxmox/test/ansible-role-proxmox/tasks/load_variables.yml': line 10, column 31, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n  # https://git.proxmox.com/?p=pve-cluster.git;a=blob;f=data/PVE/Corosync.pm;h=8b5c91e0da084da4e9ba7423176872a0c16ef5af;hb=refs/heads/stable-5#l209\n                              ^ here\n\nThere appears to be both 'k=v' shorthand syntax and YAML in this task. Only one syntax may be used.\n"
}
fatal: [192.168.15.153]: FAILED! => {
    "msg": "The conditional check 'pve_cluster_addr0 is not defined' failed. The error was: error while evaluating conditional (pve_cluster_addr0 is not defined): {{ ansible_interfaces.enp0s8.ipv4.address }}: 'list object' has no attribute 'enp0s8'\n\nThe error appears to be in '/mnt/d/proxmox/test/ansible-role-proxmox/tasks/load_variables.yml': line 10, column 31, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n  # https://git.proxmox.com/?p=pve-cluster.git;a=blob;f=data/PVE/Corosync.pm;h=8b5c91e0da084da4e9ba7423176872a0c16ef5af;hb=refs/heads/stable-5#l209\n                              ^ here\n\nThere appears to be both 'k=v' shorthand syntax and YAML in this task. Only one syntax may be used.\n"

**Comment by lae** [comment]
Oh sorry, it's been a while since I've actually used Ansible. It looks like `ansible_interfaces.enp0s8.ipv4_address` (my browser history seemed to suggest otherwise so...yeah, my mistake) wouldn't be how you access the interface's IP address, but I believe `ansible_enp0s8.ipv4.address`. So your role variable should be set like `pve_cluster_addr0: "{{ ansible_enp0s8.ipv4.address }}"`.

You can always look up the host facts of your hosts by using the setup module, like this:

```
~]$ ansible -m setup -i localhost, -c local all
[WARNING]: Platform linux on host localhost is using the discovered Python interpreter at /usr/bin/python, but future installation of another Python interpreter could change the meaning of that
path. See https://docs.ansible.com/ansible-core/2.11/reference_appendices/interpreter_discovery.html for more information.
localhost | SUCCESS => {
    "ansible_facts": {
        "ansible_all_ipv4_addresses": [
            "192.168.121.1",
            "10.199.199.10"
        ],
<cut>
        "ansible_interfaces": [
            "lo",
            "virbr1",
            "ens18"
        ],
<cut>
        "ansible_virbr1": {
            "active": false,
            "device": "virbr1",
            "interfaces": [],
            "ipv4": {
                "address": "192.168.121.1",
                "broadcast": "192.168.121.255",
                "netmask": "255.255.255.0",
                "network": "192.168.121.0"
            },
            "macaddress": "<cut>",
            "mtu": 1500,
            "promisc": false,
            "speed": -1,
            "stp": true,
            "type": "bridge"
        },
<cut>
}
```

I cut out a lot of the extraneous facts but this will basically output any information about your host that you can use in your playbooks. As we can see here, `ansible_interfaces` is actually just a list of the available interfaces on the host. The interface's networking information etc is associated with the `ansible_{{ network_interface_name }}` facts.

If you still run into a similar error after using the correct variable, let me know—I'm curious why the error message says something about the YAML in the tasks file (`ansible-lint` doesn't complain about it).
